code/game/machinery/ai_resource.dm

```diff
@@ -0,0 +1,341 @@
+
+/obj/machinery/ai_node/attack_hand(user as mob)
+	if(overheating)
+		to_chat(user, "<span class='notice'>You turn the overheating [src] off.</span>")
+		overheating = FALSE
+		update_icon(UPDATE_ICON_STATE)
+		return
+	active = !active
+	to_chat(user, "<span class = 'notice'>You turn [src] [active ? "on" : "off"].</span>")
+	if(active) // We're booting up
+		find_ai()
+		if(!assigned_ai) // No eligible AI found, abort
+			active = FALSE
+			to_chat(user, "<span class = 'warning'>ERROR: No AI detected. Shutting down...</span>")
+			to_chat(user, "<span class = 'notice'>The Processing Node turns off.</span>")
+		else // We have an AI
+			assigned_ai.program_picker.modify_resource(resource_key, resource_amount)
+			change_power_mode(ACTIVE_POWER_USE)
+	else // We're shutting down
+		if(assigned_ai)
+			assigned_ai.program_picker.modify_resource(resource_key, -resource_amount)
+			assigned_ai = null
+		change_power_mode(IDLE_POWER_USE)
+	update_icon(UPDATE_ICON_STATE)
+
+/obj/machinery/ai_node/proc/find_ai()
+	if(!resource_key)
+		return
+	for(var/mob/living/silicon/ai/new_ai as anything in GLOB.ai_list)
+		if(!assigned_ai) // Not found
+			assigned_ai = new_ai // Assign to the first AI in the list to start
+			continue
+		if(resource_key == "memory" && assigned_ai.program_picker.memory > new_ai.program_picker.memory)
+			assigned_ai = new_ai
+			continue
+		if(resource_key == "bandwidth" && assigned_ai.program_picker.bandwidth > new_ai.program_picker.bandwidth)
+			assigned_ai = new_ai
+
+/obj/machinery/ai_node/proc/overheat()
+	active = FALSE
+	if(assigned_ai)
+		assigned_ai.program_picker.modify_resource(resource_key, -resource_amount)
+		assigned_ai = null
+	obj_integrity -= (max_integrity / 3) // Overheat it three times and it breaks
+	change_power_mode(IDLE_POWER_USE)
+	update_icon(UPDATE_ICON_STATE)
+
+/obj/machinery/ai_node/proc/change_ai(mob/living/silicon/ai/new_ai)
+	if(!new_ai)
+		return
+	if(!istype(new_ai))
+		return
+	assigned_ai.program_picker.modify_resource(resource_key, -resource_amount)
+	assigned_ai = new_ai
+	assigned_ai.program_picker.modify_resource(resource_key, resource_amount)
+
+/datum/milla_safe/ai_node_process
+
+/datum/milla_safe/ai_node_process/on_run(obj/machinery/ai_node/node)
+	var/turf/simulated/L = get_turf(node)
+	if(!istype(L))
+		return
+	var/datum/gas_mixture/env = get_turf_air(L)
+	var/transfer_moles = 0.25 * env.total_moles()
+	var/datum/gas_mixture/removed = env.remove(transfer_moles)
+	if(!removed)
+		return
+	var/heat_capacity = removed.heat_capacity()
+	if(heat_capacity)
+		removed.set_temperature(removed.temperature() + node.heat_amount / heat_capacity)
+	env.merge(removed)
+	// Heat check
+	if(env.temperature() > node.overheat_temp || env.temperature() < 273) // If the temperature is outside 0-100C...
+		// Turn the node off due to temperature problems
+		node.overheat_counter++
+		if(node.overheat_counter >= 5)
+			node.overheat()
+		return
+	node.overheat_counter--
```

Comment:
is it possible that this might set overheat counter into the far negatives, meaning that it'll realistically never overheat?